### PR TITLE
Autoreg for mac-auth

### PR DESCRIPTION
**Pull request by fdurand**
# Description
When "Automatically register devices" is enabled on the portal and the connection is mac-auth then try to use authorization source to compute a
role and an access duration

# Impacts
- RADIUS authorize workflow

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Autoreg for mac-auth with an authorize source

### CODE REVIEW

File: lib/pf/role.pm [hunk 1]
```diff
@@ -395,14 +395,55 @@ sub getRegisteredRole {
...
+            }
+            if (blessed ($args->{node_info})) {
+                $args->{node_info}->merge(\%info);
+            }
+            else {
+                node_modify($args->{'mac'},%info);
+            }
+        } else {
+            $role = $args->{'node_info'}->{'category'};
+        }
```
Comment (julsemaan) [inline lib/pf/role.pm:411]: I think now that we've been copy/pasting this piece of code 3 times in that module it deserves its own sub

File: lib/pf/role.pm [hunk 2]
```diff
@@ -395,14 +395,55 @@ sub getRegisteredRole {
...
+            };
+            my $matched = pf::authentication::match2([@sources], $params);
+            $source = $matched->{source_id};
+            my $values = $matched->{values};
+            $role = $values->{$Actions::SET_ROLE};
+            my $unregdate = $values->{$Actions::SET_UNREG_DATE};
+            my $time_balance =  $values->{$Actions::SET_TIME_BALANCE};
+            my $bandwidth_balance =  $values->{$Actions::SET_BANDWIDTH_BALANCE};
+            my %info = (
+                'pid' => 'default',
```
Comment (julsemaan) [inline lib/pf/role.pm]: constant for default PID

File: lib/pf/role.pm [hunk 3]
```diff
@@ -395,14 +395,55 @@ sub getRegisteredRole {
...
-        $logger->info("Connection type is Wireless-802.11-NoEAP. Getting role from node_info" );
-        $role = $args->{'node_info'}->{'category'};
+    if ( $args->{'connection_type'} && ( ( ($args->{'connection_type'} & $WIRED_MAC_AUTH) == $WIRED_MAC_AUTH ) || ( ($args->{'connection_type'} & $WIRELESS_MAC_AUTH) == $WIRELESS_MAC_AUTH ) ) ) {
+        $logger->info("Connection type is MAC-AUTH. Getting role from node_info" );
+        my @sources = $profile->getFilteredAuthenticationSources($args->{'stripped_user_name'}, $args->{'realm'});
+        my @source = grep {uc($_->{'type'}) eq "AUTHORIZATION"} @sources;
+        if (@source) {
+            my $stripped_user = '';
+            $stripped_user = $args->{'stripped_user_name'} if(defined($args->{'stripped_user_name'}));
+            my $params = {
```
Comment (julsemaan) [inline lib/pf/role.pm]: Can we do a subroutine for building params from args that we could reuse elsewhere in this module?

File: lib/pf/role.pm [hunk 4]
```diff
@@ -732,7 +773,7 @@ sub _check_bypass {
     my $logger = get_logger();
 
     # If the role of the node is the REJECT role, then we early return as it has precedence over the bypass role and VLAN
-    if($args->{node_info}->{category} eq $REJECT_ROLE) {
+    if(defined($args->{node_info}->{category}) && $args->{node_info}->{category} eq $REJECT_ROLE) {
```
Comment (julsemaan) [inline lib/pf/role.pm:753]: Should this be a separate patch or is this related directly to this PR?

**Comment by fdurand** [comment]
Bump
